frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -51,23 +51,52 @@ type ConvertOpt struct {
 	BuildPlatforms   []specs.Platform
 }
 
-func Dockerfile2LLB(ctx context.Context, dt []byte, opt ConvertOpt) (*llb.State, *Image, error) {
-	if len(dt) == 0 {
-		return nil, nil, errors.Errorf("the Dockerfile cannot be empty")
+type platformOpt struct {
+	targetPlatform specs.Platform
+	buildPlatforms []specs.Platform
+	implicitTarget bool
+}
+
+func buildPlatformOpt(opt *ConvertOpt) *platformOpt {
+	return &platformOpt{
+		targetPlatform: *resolveTargetPlatform(opt),
+		buildPlatforms: resolveBuildPlatforms(opt),
+		implicitTarget: implicitTargetPlatform(opt),
 	}
+}
+
+func resolveBuildPlatforms(opt *ConvertOpt) []specs.Platform {
+	buildPlatforms := opt.BuildPlatforms
 
 	if opt.TargetPlatform != nil && opt.BuildPlatforms == nil {
-		opt.BuildPlatforms = []specs.Platform{*opt.TargetPlatform}
+		buildPlatforms = []specs.Platform{*opt.TargetPlatform}
 	}
-	if len(opt.BuildPlatforms) == 0 {
-		opt.BuildPlatforms = []specs.Platform{platforms.DefaultSpec()}
+	if len(buildPlatforms) == 0 {
+		buildPlatforms = []specs.Platform{platforms.DefaultSpec()}
 	}
-	implicitTargetPlatform := false
+
+	return buildPlatforms
+}
+
+func resolveTargetPlatform(opt *ConvertOpt) *specs.Platform {
 	if opt.TargetPlatform == nil {
-		implicitTargetPlatform = true
-		opt.TargetPlatform = &opt.BuildPlatforms[0]
+		return &resolveBuildPlatforms(opt)[0]
```

Comment:
nit: this `resolveBuildPlatforms()` still seems bit weird in here. I'd just do all these in `buildPlatformOpt`.